rfcs/0007-filtering.md

```diff
@@ -0,0 +1,211 @@
+
+    private static final Map<String, SubfieldSelection> eobFieldSelection = new HashMap<>();
+
+    static {
+        eobFieldSelection.put("patient", wildCard);
+        eobFieldSelection.put("provider", wildCard);
+        eobFieldSelection.put("organization", wildCard);
+        eobFieldSelection.put("facility", wildCard);
+        eobFieldSelection.put("type", wildCard);
+        eobFieldSelection.put("resourceType", wildCard);
+        eobFieldSelection.put("diagnosis", wildCard);
+        eobFieldSelection.put("procedure", wildCard);
+        eobFieldSelection.put("item", buildItemDef());
+        eobFieldSelection.put("careTeam", wildCard);
+        eobFieldSelection.put("identifier", wildCard);
+    }
+
+    private static SubfieldSelection buildItemDef() {
+        Map<String, SubfieldSelection> itemMap = new HashMap<>();
+        itemMap.put("service", wildCard);
+        itemMap.put("quantity", wildCard);
+        itemMap.put("servicedPeriod", wildCard);
+        itemMap.put("location", wildCard);
+        itemMap.put("careTeamLinkId", wildCard);
+        itemMap.put("sequence", wildCard);
+        return new SubfieldSelection(itemMap);
+    }
+    
+    // Entry point
+    public static void filterEOBs(List<ExplanationOfBenefit> eobs) {
+        for (ExplanationOfBenefit eob : eobs) {
+            trimFields(eob, eobFieldSelection);
+        }
+    }
+
+    private static void trimFields(Object objToTrim, Map<String, SubfieldSelection> fieldMapping) {
+        Field[] fields = objToTrim.getClass().getDeclaredFields();
+        for (Field field : fields) {
+            // Do not return this field, null it out.
+            if (!fieldMapping.containsKey(field.getName())) {
+                PropertyUtils.setProperty(objToTrim, field.getName(), null);
+                continue;
+            }
+
+            // Return the entire field in the results
+            SubfieldSelection selection = fieldMapping.get(field.getName());
+            if (selection.getPickMe() == WhichFields.ALL) {
+                continue;
+            }
+
+            // Return a subset of the fields of the object referenced.
+            trimFieldsThrowException(objToTrim, field, selection);
+        }
+    }
+
+    private static void trimFieldsThrowException(Object objToTrim, Field field, SubfieldSelection selection) {
+        try {
+            trimFields(field.get(objToTrim), selection.getExplictFields());
+        } catch (IllegalAccessException iae) {
+            throw new RuntimeException(iae);
+        }
+    }
+``` 
+
+### Proposed Solution: Unresolved Questions
+[Proposed Solution: Unresolved Questions]: #proposed-solution-unresolved-questions
+
+How much will this save us?
+ 
+Proposed profiling is invoking BFD with at least 10,000 EOB Requests by Patient ID (ExplanationOfBenefit?patient=) and to measure
+the aggregate time of completion of all calls.  In other words, sum up the time taken by each call over all calls.
```

Comment:
I'd strongly suggest a more formal approach towards testing and verification be proposed here, for several reasons:
1. Your early testing seemed to indicate that one library was significantly more performant than another, which leads me to worry in general that this will end up being performance-critical code.
2. Us engineers are notoriously terrible at benchmarking. One way to do it right is to treat it more like a science lab, with methods written up in advance.
3. This is definitely security-critical code and so needs to be approached with a certain amount of rigor.

Can you spend some thought and time expanding this section a bit? Specifically, I'd like to see you detail how total performance will be benchmarked and then compared by AB2D. I'd strongly suggest that the approach there ought to include hiding the initial implementation behind a feature flag, to allow for that testing. I'd also spend a little bit of time considering how best to create unit and integration tests that appropriately address the security concerns around this code.

cc @StewGoin